realm-annotations-processor/src/main/java/io/realm/processor/RealmJsonTypeHelper.java

```diff
@@ -51,12 +49,15 @@
             @Override
             public void emitTypeConversion(String setter, String fieldName, String fieldType, JavaWriter writer) throws IOException {
                 writer
-                    .emitStatement("long timestamp = json.optLong(\"%s\", -1)", fieldName)
-                    .beginControlFlow("if (timestamp > -1)")
-                        .emitStatement("obj.%s(new Date(timestamp))", setter)
+                    .beginControlFlow("if (!json.isNull(\"%s\"))", fieldName)
+                        .emitStatement("Object timestamp = json.get(\"%s\")", fieldName)
+                        .beginControlFlow("if (timestamp instanceof String)")
+                           .emitStatement("obj.%s(JsonUtils.stringToDate((String) timestamp))", setter)
+                        .nextControlFlow("else")
+                            .emitStatement("obj.%s(new Date(json.getLong(\"%s\")))", setter, fieldName)
```

Comment:
Reconsidering this. We don't check for the right type anywhere else in the JSON code. If people use the wrong type they get an exception, so this is actually intended behaviour.